### PR TITLE
[XLA:GPU] add cuDNN flash attention support in XLA

### PR DESCRIPTION
**Pull request by Cjkkkk**
Feature: 
* This PR add supports for cuDNN flash attention inference and training for `bmm-(bias)-(causalMask)-softmax-(dropout)-bmm` pattern. cuDNN flash attention is supported with cuDNN graph API and specialized runtime fusion engine, doc [here](https://docs.nvidia.com/deeplearning/cudnn/developer-guide/index.html#specialized-runtime-fusion-engines).

Benchmark:
* Verified with different PAXML GPT3 models(126M/1.3B/5B/175B) and different sharding settings (PP, DP, FSDP, TP). Benchmark showing speed up from 1.05x to 2x compared to XLA without cuDNN flash attention in A100/H100.

Convergence:
* Verified convergence with different PAXML GPT3 models(126M/5B/175B) on pile dataset with predefined configs in PAXML in A100/H100.

Others:
* Only PP with SAVE_NOTHING checkpoint policy is supported right now due to the limitation of PP implementation in PAXML.

Implementation details:
* changes of rewriter: add causal mask pattern matching and cuDNN flash attention support check(seq > 512/cuDNN >= 8.9.3).
* changes of runner/MLIR Op: 2 additional output buffers(softmax_sum/dQ_accum) and 1 additional input buffer(output from fwd) are required by cuDNN flash attention kernel. 
* changes of cuDNN graph: add flash attention fwd and bwd graph generation.
* Both thunk lowering and runtime lowering are implemented.

PR Split:
* 1st PR: SE changes only PR: https://github.com/openxla/xla/pull/6293
* 2nd PR: MLIR lowering and thunk/runtime PR: https://github.com/openxla/xla/pull/6657
* 3rd PR: Rewriter change PR: https://github.com/openxla/xla/pull/6872

### CODE REVIEW

**Comment by ddunl** [comment]
@cheshire  to review (or feel free to tag someone you think is more appropriate)

File: xla/service/gpu/cudnn_fused_mha_rewriter.cc [hunk 1]
```diff
@@ -402,6 +405,102 @@ StatusOr<bool> IsSupportedBMM2(const HloInstruction* bmm_2,
...
+      auto comp = param->parent();
+      auto mod = comp->parent();
+      auto name = comp->name();
+      auto entry_computation = mod->entry_computation();
+      bool is_causal_mask = true;
+      for (HloInstruction* instr :
+           entry_computation->MakeInstructionPostOrder()) {
+        if (instr->opcode() == HloOpcode::kWhile &&
+            instr->while_body()->name() == name) {
+          DCHECK(instr->operand(0)->opcode() == HloOpcode::kTuple);
```
Comment (cheshire) [inline xla/service/gpu/cudnn_fused_mha_rewriter.cc:484]: Prefer referring bad Status with TF_RET_CHECK, or CHECK if the error can absolutely not occur.

**Review by cheshire — CHANGES_REQUESTED** [review]

**Comment by gbaned** [comment]
Hi @Cjkkkk Can you please check @cheshire's [comments](https://github.com/openxla/xla/pull/5910#discussion_r1345497471) and keep us posted ? Thank you!

File: xla/mlir/backends/gpu/transforms/lmhlo_gpu_to_gpu_runtime.cc [hunk 2]
```diff
@@ -746,63 +748,82 @@ class FusedAttentionBackwardLowering
...
-        } else {
-          return op.emitOpError(
-              "unexpected number of operands for fused attention backward - "
-              "BMM_Bias_Mask_Softmax_Dropout_BMM");
-        }
-        break;
-
-      default:
-        return op.emitOpError("Undefined fused attention DAG signature");
+    if (!is_flash_attention) {
```
Comment (cheshire) [inline xla/mlir/backends/gpu/transforms/lmhlo_gpu_to_gpu_runtime.cc:756]: Prefer early exit to avoid excessive nesting?

File: xla/service/gpu/backend_configs.proto [hunk 3]
```diff
@@ -161,4 +161,7 @@ message CudnnfMHABackendConfig {
 
   // Random seed used by dropout
   int64 seed = 15;
+
+  // Is flash attention
+  bool is_flash_attention = 20;
```
Comment (cheshire) [inline xla/service/gpu/backend_configs.proto:166]: Is this parameter composable with all the others? E.g. can you have a valid flash attention with any dot dimension numbers/dropout rate/fmha scale/seed/etc?

File: xla/service/gpu/cudnn_fused_mha_rewriter.cc [hunk 4]
```diff
@@ -402,6 +405,102 @@ StatusOr<bool> IsSupportedBMM2(const HloInstruction* bmm_2,
   return true;
 }
 
+StatusOr<bool> isFlashAttention(HloInstruction* bmm_1, bool is_causal_mask,
```
Comment (cheshire) [inline xla/service/gpu/cudnn_fused_mha_rewriter.cc:408]: Nit: `IsFlashAttention`

File: xla/service/gpu/cudnn_fused_mha_rewriter.cc [hunk 5]
```diff
@@ -402,6 +405,102 @@ StatusOr<bool> IsSupportedBMM2(const HloInstruction* bmm_2,
...
+                                std::string& custom_call_name) {
```
Comment (cheshire) [inline xla/service/gpu/cudnn_fused_mha_rewriter.cc:409]: Nit: absl::string_view

File: xla/service/gpu/cudnn_fused_mha_rewriter.cc [hunk 6]
```diff
@@ -63,10 +63,13 @@ struct MatchFwdResult {
   HloInstruction* matched_mask = nullptr;
   HloInstruction* matched_scale = nullptr;
   HloInstruction* matched_softmax_input = nullptr;
+  HloInstruction* matched_reduce_sum = nullptr;
 
   double matched_dropout_rate = 0.0;
   bool need_canonicalization = false;
   bool is_training = false;
+  bool is_causal_mask = false;
```
Comment (cheshire) [inline xla/service/gpu/cudnn_fused_mha_rewriter.cc:71]: Would be great to add a short comment describing semantics of those attributes.

File: xla/stream_executor/cuda/cuda_dnn.cc [hunk 7]
```diff
@@ -3604,12 +3604,16 @@ enum CudnnfMHAUid {
   MASK_ID,
   ZERO_VAL_ID,
   ONE_VAL_ID,
+  NEG_INFINITY_ID,
```
Comment (cheshire) [inline xla/stream_executor/cuda/cuda_dnn.cc:3607]: Is it possible to factor out SE changes into a separate PR?

**Review by cheshire — CHANGES_REQUESTED** [review]

File: xla/stream_executor/cuda/cuda_dnn.cc (hunk 7 above)
Comment (Cjkkkk) [reply xla/stream_executor/cuda/cuda_dnn.cc:3607]: Sure. Let me get one. So we keep 2 PR one for SE and one for the rest?

Comment (cheshire) [reply xla/stream_executor/cuda/cuda_dnn.cc:3607]: Generally the more the better, as long as they are self-contained. So at least 2. Not sure if it's possible to subdivide it further.

File: xla/service/gpu/backend_configs.proto (hunk 3 above)
Comment (Cjkkkk) [reply xla/service/gpu/backend_configs.proto:166]: flash attention is valid with any dot dumensions/dropout rate/fmha scale/seed. The constraint is that head dim should be divisible by 8 and up to 128, sequence length Q and K should be equal.

File: xla/service/gpu/cudnn_fused_mha_rewriter.cc (hunk 5 above)
Comment (Cjkkkk) [reply xla/service/gpu/cudnn_fused_mha_rewriter.cc:409]: changed to string_view

File: xla/service/gpu/cudnn_fused_mha_rewriter.cc (hunk 4 above)
Comment (Cjkkkk) [reply xla/service/gpu/cudnn_fused_mha_rewriter.cc:408]: changed to `IsFlashAttention`

File: xla/service/gpu/cudnn_fused_mha_rewriter.cc [hunk 8]
```diff
@@ -402,6 +405,102 @@ StatusOr<bool> IsSupportedBMM2(const HloInstruction* bmm_2,
...
+  // flash attention TODO: check contracting dim here
+  std::vector<int64_t> contracting_dims_bmm1 = GetDimensionVector(
+      bmm_1->operand(0)->shape().dimensions(),
+      bmm_1->dot_dimension_numbers().lhs_contracting_dimensions());
+  // for now, seq_q and seq_k should be equal for flash attention to work
+  // flash attention only supports fixed topology so we check if custom call is
+  // such topology by checking custom_call_name
+  TF_RET_CHECK(seq_q.size() == 1);
+  TF_RET_CHECK(seq_k.size() == 1);
+  auto is_fixed_topology = [=]() {
```
[the remark this answers is not in the document]
Comment (Cjkkkk) [reply xla/service/gpu/cudnn_fused_mha_rewriter.cc:439]: fixed

File: xla/service/gpu/cudnn_fused_mha_rewriter.cc [hunk 9]
```diff
@@ -402,6 +405,102 @@ StatusOr<bool> IsSupportedBMM2(const HloInstruction* bmm_2,
...
+          bmm_1->dot_dimension_numbers().rhs_batch_dimensions(),
+          bmm_1->dot_dimension_numbers().rhs_contracting_dimensions()));
+
+  std::vector<int64_t> seq_q =
+      GetDimensionVector(bmm_1->operand(0)->shape().dimensions(), seq_q_dims);
+
+  std::vector<int64_t> seq_k =
+      GetDimensionVector(bmm_1->operand(1)->shape().dimensions(), seq_k_dims);
+
+  // flash attention TODO: check contracting dim here
```
[the remark this answers is not in the document]
Comment (Cjkkkk) [reply xla/service/gpu/cudnn_fused_mha_rewriter.cc:430]: fixed

File: xla/service/gpu/cudnn_fused_mha_rewriter.cc [hunk 10]
```diff
@@ -402,6 +405,102 @@ StatusOr<bool> IsSupportedBMM2(const HloInstruction* bmm_2,
...
+              m::Broadcast(m::Constant()))))))));
+  HloInstruction* param = nullptr;
+  HloInstruction* gte = nullptr;
+  auto causal_mask_pattern_fwd =
+      m::Broadcast(m::Bitcast(m::GetTupleElement(&gte, m::Parameter(&param))));
+  auto causal_mask_pattern = m::AnyOf<HloInstruction>(
+      causal_mask_pattern_fwd_remat, causal_mask_pattern_fwd,
+      causal_mask_pattern_bwd);
+  if (Match(mask, causal_mask_pattern)) {
+    if (param != nullptr) {
```
[the remark this answers is not in the document]
Comment (Cjkkkk) [reply xla/service/gpu/cudnn_fused_mha_rewriter.cc:471]: `causal_mask_pattern` contains a `m::AnyOf<HloInstruction>` so `param` might be empty.

File: xla/mlir/backends/gpu/transforms/lmhlo_gpu_to_gpu_runtime.cc (hunk 2 above)
Comment (Cjkkkk) [reply xla/mlir/backends/gpu/transforms/lmhlo_gpu_to_gpu_runtime.cc:756]: Not sure if I misunderstood your question here. But we need to dispatch between flash attention and regular fused attention here so there is not early exit.

**Comment by wangsiyu** [comment]
@Cjkkkk  Hi, this PR looks useful, but I try to run the unit test of flash attention and get CUDNN_STATUS_NOT_SUPPORTED error. Does it needs the specified environment? My docker is based on cuda 11.8 and cudnn 8.9.0

**Comment by Cjkkkk** [comment]
> @Cjkkkk Hi, this PR looks useful, but I try to run the unit test of flash attention and get CUDNN_STATUS_NOT_SUPPORTED error. Does it needs the specified environment? My docker is based on cuda 11.8 and cudnn 8.9.0

It would be best to use cudnn 8.9.5 here. Also it would be helpful to debug if you can use a debug version and set `export CUDNN_PRINT_DBG=1` to get the exact reason for `CUDNN_STATUS_NOT_SUPPORTED `.

**Comment by Cjkkkk** [comment]
@cheshire I added first PR with only SE changes https://github.com/openxla/xla/pull/6293. We can review that PR first. Any thoughts on how to split the rest so the review process can go more smoothly?

**Comment by Cjkkkk** [comment]
Also I dont quite understand what is this failure: `Buildifier / buildifier-lint (pull_request_target) `. How do I resolve this issue? @ddunl

**Comment by Cjkkkk** [comment]
> @cheshire I added first PR with only SE changes #6293. We can review that PR first. Any thoughts on how to split the rest so the review process can go more smoothly?

With only SE changes, the compilation would fail and it requires thunk/runner to work. Maybe rewriter/runner/MLIR op as 1st PR and SE changes as the 2nd PR. Both would have around 1.5k LOC.

**Comment by ddunl** [comment]
Oh woops, I evidently configured that wrong. I'll disable and reenable later, sorry for false alarm

File: xla/service/gpu/cudnn_fused_mha_rewriter_test.cc [hunk 11]
```diff
@@ -2695,6 +2695,580 @@ ENTRY main.82 {
...
+  broadcast.26 = bf16[2,6,2048,2048]{3,2,1,0} broadcast(divide.3), dimensions={0,1,2}
+  multiply.4 = bf16[2,6,2048,2048]{3,2,1,0} multiply(dot.60, broadcast.26)
+  multiply.5 = bf16[2,6,2048,2048]{3,2,1,0} multiply(multiply.4, exponential.1)
+  reduce.70 = bf16[2,6,2048]{2,1,0} reduce(multiply.5, constant.16), dimensions={3}, to_apply=region_2.66
+  negate.2 = bf16[2,6,2048]{2,1,0} negate(reduce.70)
+  broadcast.31 = bf16[2,6,2048,2048]{3,2,1,0} broadcast(negate.2), dimensions={0,1,2}
+  add.5 = bf16[2,6,2048,2048]{3,2,1,0} add(divide.4, broadcast.31)
+  multiply.8 = bf16[2,6,2048,2048]{3,2,1,0} multiply(add.5, exponential.1)
+  multiply.9 = bf16[2,6,2048,2048]{3,2,1,0} multiply(multiply.8, broadcast.29)
+  dot.90 = bf16[2,6,2048,128]{3,2,1,0} dot(multiply.9, Arg_1.2), lhs_batch_dims={0,1}, lhs_contracting_dims={3}, rhs_batch_dims={0,1}, rhs_contracting_dims={3}
```
[the remark this answers is not in the document]
Comment (Cjkkkk) [reply xla/service/gpu/cudnn_fused_mha_rewriter_test.cc:2768]: From my experience, maybe we could do this for softmax patterns but attention patterns are very complex and different.

**Comment by wangsiyu** [comment]
> > @Cjkkkk Hi, this PR looks useful, but I try to run the unit test of flash attention and get CUDNN_STATUS_NOT_SUPPORTED error. Does it needs the specified environment? My docker is based on cuda 11.8 and cudnn 8.9.0
> 
> It would be best to use cudnn 8.9.5 here. Also it would be helpful to debug if you can use a cuDNN debug version and set `export CUDNN_PRINT_DBG=1` to get the exact reason for `CUDNN_STATUS_NOT_SUPPORTED `. And HLO would be useful as well.

Thanks for reply. Another question is the unit test is only for forward phase. Is the backward valid in your test ?

**Comment by Cjkkkk** [comment]
> > With only SE changes, the compilation would fail and it requires thunk/runner to work
> 
> Could you elaborate? Is it possible to have somewhat self-contained SE change?

I just talked with Tj on this, will only include cudnn graph for the first PR. Will update today with a self-contained changes.

**Comment by Cjkkkk** [comment]
> > > @Cjkkkk Hi, this PR looks useful, but I try to run the unit test of flash attention and get CUDNN_STATUS_NOT_SUPPORTED error. Does it needs the specified environment? My docker is based on cuda 11.8 and cudnn 8.9.0
> > 
> > 
> > It would be best to use cudnn 8.9.5 here. Also it would be helpful to debug if you can use a cuDNN debug version and set `export CUDNN_PRINT_DBG=1` to get the exact reason for `CUDNN_STATUS_NOT_SUPPORTED `. And HLO would be useful as well.
> 
> Thanks for reply. Another question is the unit test is only for forward phase. Is the backward valid in your test ?

backward is also valid. We are adding more tests as we speak.

**Comment by tdanyluk** [comment]
PR rotation:
Hi @Cjkkkk, I think that this PR is waiting for an action on your side. Please let us know if you need anything from Google's side.

**Comment by Cjkkkk** [comment]
> PR rotation: Hi @Cjkkkk, I think that this PR is waiting for an action on your side. Please let us know if you need anything from Google's side.

This PR is just for reference and maintain a corresponding version between our internal branch. We split this PR into 3 PR. The first one https://github.com/openxla/xla/pull/6293 is approved so we can move on to create the rest 2 PR.

**Comment by Cjkkkk** [comment]
@cheshire This is the 2nd PR https://github.com/openxla/xla/pull/6657 following 1st PR with MLIR lowering and thunk/runtime. The last PR will contain rewriter/transpose fusion/tests.

**Comment by Cjkkkk** [comment]
No need to track this PR: status is updated in this split PR: https://github.com/openxla/xla/pull/6872.

**Comment by kamaljeeti** [comment]
Hi @Cjkkkk, 
Please confirm can I close this PR as it is not required so. Thanks.

**Comment by kamaljeeti** [comment]
Hi @Cjkkkk, 
Closing this for now, as it is not required to track. Thanks.
